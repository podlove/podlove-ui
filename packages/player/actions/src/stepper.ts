import { createAction } from 'redux-actions'
import { STEP_FORWARD, STEP_BACKWARDS } from './types.js'

export const stepForward = createAction<void>(STEP_FORWARD)
export const stepBackwards = createAction<void>(STEP_BACKWARDS)
