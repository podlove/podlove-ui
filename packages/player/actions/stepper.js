import { createAction } from './action-creator'
import { STEP_FORWARD, STEP_BACKWARDS } from './types'

export const stepForward = createAction(STEP_FORWARD)
export const stepBackwards = createAction(STEP_BACKWARDS)
