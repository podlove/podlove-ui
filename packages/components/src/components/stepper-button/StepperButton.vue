<template>
  <button
    :id="`stepper-button--${type}`"
    class="stepper-button opacity-100 hover:opacity-75"
    @click="clickHandler"
  >
    <icon :type="type" :color="color" />
    <slot />
  </button>
</template>

<script>
import { stepForward, stepBackwards } from '@podlove/player-actions/stepper'
import Icon from '../icons/Icon'
import { color } from '../../defaults'

export default {
  components: {
    Icon
  },
  props: {
    type: {
      type: String,
      required: true,
      validator: (val) => ['forward', 'backwards'].includes(val)
    },
    color: {
      type: String,
      default: color
    }
  },
  emit: {
    forward: null,
    backwards: null
  },
  methods: {
    clickHandler() {
      switch (this.type) {
        case 'forward':
          this.$emit('forward', stepForward())
          break
        case 'backwards':
          this.$emit('backwards', stepBackwards())
          break
      }
    }
  }
}
</script>

<style lang="scss" scoped></style>
