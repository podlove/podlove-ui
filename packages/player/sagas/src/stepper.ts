import { takeEvery, select, put } from 'redux-saga/effects';
import { min, max } from 'lodash-es';
import { requestPlaytime } from '@podlove/player-actions/timepiece';
import * as stepper from '@podlove/player-actions/stepper';

export const stepperSaga = ({
  selectDuration,
  selectPlaytime,
  selectLivesync
}: {
  selectDuration: (input: any) => number;
  selectPlaytime: (input: any) => number;
  selectLivesync: (input: any) => number;
}) =>
  function* saga() {
    yield takeEvery(stepper.stepForward.toString(), stepForward, {
      selectDuration,
      selectPlaytime,
      selectLivesync
    });
    yield takeEvery(stepper.stepBackwards.toString(), stepBackward, { selectPlaytime });
  };

export function* stepForward({
  selectDuration,
  selectPlaytime,
  selectLivesync
}: {
  selectDuration: (input: any) => number;
  selectPlaytime: (input: any) => number;
  selectLivesync: (input: any) => number;
}) {
  const duration = yield select(selectDuration);
  const playtime = yield select(selectPlaytime);
  const livesync = yield select(selectLivesync);

  const upperLimit = livesync > 0 ? livesync : duration;

  const targetPlaytime = min([playtime + 30 * 1000, upperLimit]);
  yield put(requestPlaytime(targetPlaytime));
}

export function* stepBackward({ selectPlaytime }: { selectPlaytime: (input: any) => number }) {
  const playtime = yield select(selectPlaytime);
  const targetPlaytime = max([playtime - 15 * 1000, 0]);
  yield put(requestPlaytime(targetPlaytime));
}
