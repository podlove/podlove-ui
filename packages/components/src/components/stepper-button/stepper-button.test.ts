import { mount } from '@vue/test-utils';
import { stepForward, stepBackwards } from '@podlove/player-actions/stepper';
import { describe, test, expect } from 'vitest';

import StepperButton from './StepperButton.vue';

describe('StepperButton', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(StepperButton, {
        props: {
          type: 'forward'
        }
      });

      expect(wrapper.vm).toBeTruthy();
    });
  });

  describe('props', () => {
    const types = ['forward', 'backwards'];

    types.forEach((type) => {
      test(`should render for rype ${type}`, async () => {
        const wrapper = await mount(StepperButton, {
          props: {
            type,
            color: '#fff'
          }
        });

        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });

  describe('events', () => {
    test('should emit STEP_FORWARD on type forward click', async () => {
      const wrapper = await mount(StepperButton, {
        props: {
          type: 'forward'
        }
      });

      wrapper.trigger('click');

      expect(wrapper.emitted().forward).toEqual([[stepForward()]]);
    });

    test('should emit STE_BACKWARD on type backward click', async () => {
      const wrapper = await mount(StepperButton, {
        props: {
          type: 'backwards'
        }
      });

      wrapper.trigger('click');

      expect(wrapper.emitted().backwards).toEqual([[stepBackwards()]]);
    });
  });
});
