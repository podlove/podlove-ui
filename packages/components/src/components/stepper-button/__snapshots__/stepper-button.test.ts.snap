// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StepperButton > props > should render for rype backwards 1`] = `
<button
  class="podlove-component-stepper-button opacity-100 hover:opacity-75"
  color="#fff"
  data-test="podlove-component-stepper-button--backwards"
  data-v-22b488d6=""
>
  <!--v-if-->
  <svg
    aria-hidden="true"
    class="podlove-component-icon"
    data-v-22b488d6=""
    data-v-258dccf6=""
    fill="none"
    height="24"
    viewBox="0 0 32 24"
    width="31.92"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    <path
      d="M5.65 3.5C11.8333 3.5 12.8167 3.5 19 3.5C24.2467 3.5 28.5 7.75329 28.5 13C28.5 18.2467 24.2467 22.5 19 22.5H15"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.3"
    />
    <path
      d="M2.64507 2.93557L4.75612 1.08839C5.24106 0.664074 6 1.00846 6 1.65283V5.34717C6 5.99154 5.24106 6.33593 4.75612 5.91161L2.64507 4.06443C2.30357 3.76562 2.30357 3.23438 2.64507 2.93557Z"
      fill="currentColor"
    />
    <path
      d="M10.1245 15.62V17H15.4165V15.62H13.7605V9.044H12.2725L10.0765 9.944L10.4845 11.156L12.0445 10.52V15.62H10.1245ZM16.309 13.508C16.597 13.46 17.449 13.424 18.001 13.424C19.441 13.424 20.137 13.736 20.137 14.66C20.137 15.452 19.693 15.86 18.901 15.86C18.073 15.86 17.689 15.476 17.773 14.684H16.045C15.781 16.364 16.633 17.156 18.757 17.156C20.917 17.156 21.901 16.124 21.901 14.528C21.901 12.788 20.797 12.14 18.685 12.14C18.433 12.14 18.157 12.152 17.905 12.164L18.037 10.544H21.349L21.481 9.044H16.597L16.309 13.508Z"
      fill="currentColor"
    />
    
  </svg>
  
  
</button>
`;

exports[`StepperButton > props > should render for rype forward 1`] = `
<button
  class="podlove-component-stepper-button opacity-100 hover:opacity-75"
  color="#fff"
  data-test="podlove-component-stepper-button--forward"
  data-v-22b488d6=""
>
  <svg
    aria-hidden="true"
    class="podlove-component-icon"
    data-v-22b488d6=""
    data-v-258dccf6=""
    fill="none"
    height="24"
    viewBox="0 0 32 24"
    width="31.92"
    xmlns="http://www.w3.org/2000/svg"
  >
    
    <path
      d="M26.35 3.5C20.1667 3.5 19.1833 3.5 13 3.5C7.75329 3.5 3.5 7.75329 3.5 13C3.5 18.2467 7.75329 22.5 13 22.5H17"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="1.3"
    />
    <path
      d="M29.3549 2.93557L27.2439 1.08839C26.7589 0.664074 26 1.00846 26 1.65283V5.34717C26 5.99154 26.7589 6.33593 27.2439 5.91161L29.3549 4.06443C29.6964 3.76562 29.6964 3.23438 29.3549 2.93557Z"
      fill="currentColor"
    />
    <path
      d="M11.9726 13.604C13.1366 13.604 13.4606 14.036 13.4606 14.732C13.4606 15.368 13.1006 15.86 12.1646 15.86C11.2766 15.86 10.8686 15.488 10.9406 14.672H9.20064C8.96064 16.376 9.83664 17.156 12.0686 17.156C14.2406 17.156 15.2246 16.184 15.2246 14.72C15.2246 13.532 14.5886 12.944 13.4966 12.8V12.776C14.3726 12.548 14.9846 11.888 14.9846 10.952C14.9846 9.716 14.2646 8.888 12.2246 8.888C10.3046 8.888 9.21264 9.68 9.28464 11.288H10.9886C10.9646 10.652 11.2526 10.172 12.1166 10.172C12.8606 10.172 13.2206 10.52 13.2206 11.204C13.2206 11.816 12.8726 12.26 11.9726 12.26H10.8806V13.604H11.9726ZM19.2417 17.156C21.1977 17.156 22.5057 15.968 22.5057 12.908C22.5057 9.98 21.4257 8.888 19.3737 8.888C17.4057 8.888 16.0977 10.064 16.0977 13.124C16.0977 16.052 17.1897 17.156 19.2417 17.156ZM19.2897 15.824C18.3657 15.824 17.8737 15.248 17.8737 13.04C17.8737 10.856 18.4017 10.22 19.3137 10.22C20.2377 10.22 20.7417 10.796 20.7417 12.992C20.7417 15.2 20.2017 15.824 19.2897 15.824Z"
      fill="currentColor"
    />
    
  </svg>
  <!--v-if-->
  
  
</button>
`;
